File: apps/ios/LightBeam/Services/CanonicalCbor.swift
import Foundation

/// Deterministic CBOR encoder (sorted string keys) matching web `canonicalCbor.ts`.
enum CanonicalCbor {
    static func encode(_ value: Any?) -> Data {
        var out = Data()
        write(&out, value)
        return out
    }

    private static func write(_ out: inout Data, _ value: Any?) {
        guard let value else {
            out.append(0xF6)
            return
        }
        switch value {
        case let b as Bool:
            out.append(b ? 0xF5 : 0xF4)
        case let n as Int:
            writeInt(&out, n)
        case let n as Int64:
            writeInt(&out, Int(clamping: n))
        case let n as UInt64:
            writeTypeLen(&out, major: 0, len: Int(clamping: n))
        case let n as NSNumber:
            // Avoid treating Bool as NSNumber (Bool bridges to NSNumber).
            if CFGetTypeID(n) == CFBooleanGetTypeID() {
                out.append(n.boolValue ? 0xF5 : 0xF4)
            } else {
                writeInt(&out, n.intValue)
            }
        case let s as String:
            let bytes = Array(s.utf8)
            writeTypeLen(&out, major: 3, len: bytes.count)
            out.append(contentsOf: bytes)
        case let arr as [Any]:
            writeTypeLen(&out, major: 4, len: arr.count)
            for item in arr { write(&out, item) }
        case let dict as [String: Any]:
            let keys = dict.keys.sorted()
            writeTypeLen(&out, major: 5, len: keys.count)
            for k in keys {
                write(&out, k)
                write(&out, dict[k] ?? NSNull())
            }
        case is NSNull:
            out.append(0xF6)
        default:
            out.append(0xF6)
        }
    }

    private static func writeInt(_ out: inout Data, _ n: Int) {
        if n >= 0 {
            writeTypeLen(&out, major: 0, len: n)
        } else {
            writeTypeLen(&out, major: 1, len: -1 - n)
        }
    }

    private static func writeTypeLen(_ out: inout Data, major: Int, len: Int) {
        let mt = UInt8(major << 5)
        if len < 24 {
            out.append(mt | UInt8(len))
        } else if len < 0x100 {
            out.append(mt | 24)
            out.append(UInt8(len))
        } else if len < 0x10000 {
            out.append(mt | 25)
            out.append(UInt8((len >> 8) & 0xFF))
            out.append(UInt8(len & 0xFF))
        } else {
            let u = UInt32(len)
            out.append(mt | 26)
            out.append(UInt8((u >> 24) & 0xFF))
            out.append(UInt8((u >> 16) & 0xFF))
            out.append(UInt8((u >> 8) & 0xFF))
            out.append(UInt8(u & 0xFF))
        }
    }
}

File: apps/android/app/src/main/java/com/goldmann/lightbeam/protocol/CanonicalCbor.kt
package com.goldmann.lightbeam.protocol

/** Deterministic CBOR (sorted string keys) matching web canonicalCbor.ts. */
object CanonicalCbor {
    fun encode(value: Any?): ByteArray {
        val out = ArrayList<Byte>()
        write(out, value)
        return out.toByteArray()
    }

    private fun write(out: MutableList<Byte>, value: Any?) {
        when (value) {
            null -> out.add(0xf6.toByte())
            is Boolean -> out.add(if (value) 0xf5.toByte() else 0xf4.toByte())
            is Number -> writeInt(out, value.toLong())
            is String -> {
                val bytes = value.toByteArray(Charsets.UTF_8)
                writeTypeLen(out, 3, bytes.size.toLong())
                bytes.forEach { out.add(it) }
            }
            is List<*> -> {
                writeTypeLen(out, 4, value.size.toLong())
                value.forEach { write(out, it) }
            }
            is Map<*, *> -> {
                val keys = value.keys.map { it.toString() }.sorted()
                writeTypeLen(out, 5, keys.size.toLong())
                for (k in keys) {
                    write(out, k)
                    write(out, value[k])
                }
            }
            else -> out.add(0xf6.toByte())
        }
    }

    private fun writeInt(out: MutableList<Byte>, n: Long) {
        if (n >= 0) writeTypeLen(out, 0, n)
        else writeTypeLen(out, 1, -1 - n)
    }

    private fun writeTypeLen(out: MutableList<Byte>, major: Int, len: Long) {
        val mt = (major shl 5) and 0xFF
        when {
            len < 24 -> out.add((mt or len.toInt()).toByte())
            len < 0x100 -> {
                out.add((mt or 24).toByte())
                out.add(len.toByte())
            }
            len < 0x10000 -> {
                out.add((mt or 25).toByte())
                out.add(((len shr 8) and 0xFF).toByte())
                out.add((len and 0xFF).toByte())
            }
            else -> {
                out.add((mt or 26).toByte())
                out.add(((len ushr 24) and 0xFF).toByte())
                out.add(((len ushr 16) and 0xFF).toByte())
                out.add(((len ushr 8) and 0xFF).toByte())
                out.add((len and 0xFF).toByte())
            }
        }
    }
}
